### ISSUE TITLE
Global constants

### ISSUE THREAD

**Issue by mzuenni** [Kattis/problem-package-format#163]
I would propose the following changes to global constants:

- Drop the format specifier
  - The formats are very latex specific but latex is already capable of formatting numbers in the proposed ways (and even more) see for example the `siunitx` package
  - `s` and `c` are not really well defined for floating-point numbers since you would want a decimal representation for these to make sense but then you should also specify the number of decimals.
  - overall this makes stuff a lot easier
- Change the regex for constants
  - `X` is just really weird and files with it look much more confusing than more standard replacement tokens like `{%blub%}`
  - We should allow more characters for the name of the sequence, I would propose `[a-zA-Z0-9_.-]+` which would allow stuff like `n_min` and `n_max` as names which would make clearer what the intent of these constants is.
- Automatically add a few constants (and reserve their names as constants):
  - For submissions:
    - `timelimit`: for example, this allows one to write a submission which tries "random" solutions as long as it has time
  - For statements:
    - `timelimit`: this is often displayed anyway and by this, we only need one mechanism to insert additional information
    - `problemauthor`: same as above.
    - `problemyamlname`: same as above (this is the name already used by BAPCtool's)
  - Also for statements but these might be too `BAPCtools` specific...
    - `problemlabel`: for contests, we typically assign a letter to each problem
    - `problemdir`: the path to the problem
    - `problemdirname`: the folder name of the problem
    - `builddir`: the path where the statement is actually build by latex.
  - maybe we want to reserve all constants starting with `problem`?

cc @RagnarGrootKoerkamp

**Comment by RagnarGrootKoerkamp** [comment]
Some comments:
- I'd be happy to simplify things and drop the format specifiers. The question is whether HTML rendering works well without them. We could also do without them for now and add them back later.
- The reason I was pushing the original `X..X` syntax is so that it's a valid identifier in most languages. Once we move to `{%blah%}`, that's anyway not valid and this concern goes away. The main question here is whether autoformatting of source code doesn't mess up `{%iden-ti.fier%}`. If it does, that would be quite annoying...

**Comment by niemela** [comment]
> * Drop the format specifier

I would be fine with this, and as @RagnarGrootKoerkamp pointed out, if it turns out we really need it, it could be added later.

> * Change the regex for constants
>     * `X` is just really weird and files with it look much more confusing than more standard replacement tokens like `{%blub%}`

Yes, it *is* weird (and, yes, it looks really out of place pretty much wherever you use it). 

>> * The main question here is whether autoformatting of source code doesn't mess up `{%iden-ti.fier%}`. If it does, that would be quite annoying...

In most autoformatting I've seen it would bot mess up, at least no more than the internals of `{%iden-ti.fier%}` which IMO would not be *that* bad.

If *that* is the main question, then I think this would be an improvement. Especially if we also drop the format specifier.

> * [...]
>   * We should allow more characters for the name of the sequence, I would propose `[a-zA-Z0-9_.-]+` which would allow stuff like `n_min` and `n_max` as names which would make clearer what the intent of these constants is.

That makes sense.

> * Automatically add a few constants (and reserve their names as constants)

That's a nice idea.

> * For submissions:
>   
>   * `timelimit`: for example, this allows one to write a submission which tries "random" solutions as long as it has time

The time limit would typically not be known at preprocessing time (i.e. at the very beginning). Even if we could solve that, at least some submission (AC and TLE) would have to be excluded because they are used to determine (or verify) the time limit.

I would prefer not adding this.

> * For statements:
>   
>   * `timelimit`: this is often displayed anyway and by this, we only need one mechanism to insert additional information

But this is displayed and added outside of the problem package. In fact, the time limit is typically not included in the problem statement snippet in the problem package, and IMO that is how it should be.

>   * `problemauthor`: same as above.

This is set in problem.yaml. Where would you need it as a variable (in the problem package)?

>   * `problemyamlname`: same as above (this is the name already used by BAPCtool's)

Could you elaborate on why this is useful? In LaTeX we can use `\problemname{name}`, but maybe there are other needs?

> * Also for statements but these might be too `BAPCtools` specific...
>   
>   * `problemlabel`: for contests, we typically assign a letter to each problem

Same comment as for time limit. The problem label is not a property of the problem package, but of the problem set package (that has not yet been specified).

>   * `problemdir`: the path to the problem

Where would you use that? Wouldn't this just encourage you to write problem packages that would break when moved?

>   * `problemdirname`: the folder name of the problem

Same as above.

>   * `builddir`: the path where the statement is actually build by latex.

Same as above, but maybe to a lesser extent?
 
> * maybe we want to reserve all constants starting with `problem`?

**Comment by mzuenni** [comment]
> > problemyamlname: same as above (this is the name already used by BAPCtool's)

> Could you elaborate on why this is useful? In LaTeX we can use \problemname{name}, but maybe there are other needs?

Yes in Latex we need to write `\problemname{<actual name here>}` right now but wouldn't it be nice to be able to just write `\problemname{{%problemyamlname%}}`? This reduces the chance of typos and other mismatches between the yaml name and the one displayed in the statement. IMO only in cases where you want some different formatting of the title you should write the name there.

The other constants I mentioned are somewhat BAPCtools specific. To build the actual problem statement BAPCtools has some latex template that is added to the `problem_statement/problem.en.tex` to get a full latex file and that template contains the mentioned constants.

**Comment by niemela** [comment]
> > > problemyamlname: same as above (this is the name already used by BAPCtool's)
> 
> > Could you elaborate on why this is useful? In LaTeX we can use \problemname{name}, but maybe there are other needs?
> 
> Yes in Latex we need to write `\problemname{<actual name here>}` right now but wouldn't it be nice to be able to just write `\problemname{{%problemyamlname%}}`? This reduces the chance of typos and other mismatches between the yaml name and the one displayed in the statement. 

Right, but as currently specified `\problemname{}` means exactly `\problemname{{%problemyamlname%}}`, so you definitely don't need it for that.

Of course, we have not defined anything like that for Markdown, and I think maybe we should?

> IMO only in cases where you want some different formatting of the title you should write the name there.

Agreed, that was the idea.

> The other constants I mentioned are somewhat BAPCtools specific. To build the actual problem statement BAPCtools has some latex template that is added to the `problem_statement/problem.en.tex` to get a full latex file and that template contains the mentioned constants.

Ok, I think those would all make perfect sense in a *problem set* package, and arguably that is how BAPCtools uses them. I would not want to add them to the *problem* package.

**Comment by thorehusfeldt** [comment]
For the record, I added som code to BAPCtools that makes sure that problemyamlname and the title in \problemname, if both specified, actually agree. (Also across natural languages.) But this is quite clumsy.

https://github.com/RagnarGrootKoerkamp/BAPCtools/commit/f718d204e6949b1b605030a6b640ab60a92682aa#diff-05722462d9bf149053eac334e5bc71f43b658c8c82d124db87547d9d453a9e84R74
